File: playstead-server/lib/playstead/availability_vocabulary.ex
defmodule Playstead.AvailabilityVocabulary do
  @moduledoc """
  The frozen, six-value console availability filter vocabulary
  (plan 03-13, LIBR-02 gap closure). These are the UI-SPEC's six
  card-eligible ladder states with pinned/verified merged into one
  `ready_offline` value (`03-UI-SPEC.md` "Search and filters"; D-13).
  `safe_to_evict` is deliberately excluded — it is a storage-view
  concept, never a console filter chip.

  `shared/availability-vocabulary.json` is this module's test-fixture
  mirror (the `Playstead.SaveVocabulary` pattern): a JSON file read only
  by `availability_vocabulary_contract_test.exs` to prove the two never
  drift apart. No shipped runtime code path reads that JSON file.

  `valid?/1` is a plain membership test over a frozen list of binaries.
  It must never dynamically convert a client-supplied value into an
  atom (T-03-13-01) — unbounded atom creation from untrusted input
  exhausts the non-garbage-collected atom table.
  """

  @values [
    "needs_attention",
    "missing_dependency",
    "downloading",
    "ready_offline",
    "queued",
    "server_only"
  ]

  @labels %{
    "needs_attention" => "Needs attention",
    "missing_dependency" => "Missing dependency",
    "downloading" => "Downloading",
    "ready_offline" => "Ready offline",
    "queued" => "Queued",
    "server_only" => "On server"
  }

  @accessible_names %{
    "needs_attention" => "Filter by needs attention",
    "missing_dependency" => "Filter by missing dependency",
    "downloading" => "Filter by downloading",
    "ready_offline" => "Filter by ready offline",
    "queued" => "Filter by queued",
    "server_only" => "Filter by on server"
  }

  @doc "The six frozen filter values, in display order."
  @spec values() :: [String.t()]
  def values, do: @values

  @doc "Membership test over the frozen vocabulary — never an atom conversion."
  @spec valid?(term()) :: boolean()
  def valid?(value) when is_binary(value), do: value in @values
  def valid?(_value), do: false

  @doc "The visible chip label for `value`, or `nil` if not in the vocabulary."
  @spec label(String.t()) :: String.t() | nil
  def label(value), do: Map.get(@labels, value)

  @doc "The accessible name (`aria-label`) for `value`, or `nil` if not in the vocabulary."
  @spec accessible_name(String.t()) :: String.t() | nil
  def accessible_name(value), do: Map.get(@accessible_names, value)
end

File: playstead-mac/Playstead/Cache/AvailabilityVocabulary.swift
import Foundation

/// The frozen, six-value console availability filter vocabulary
/// (plan 03-13/03-14, LIBR-02 gap closure), mirroring
/// `Playstead.AvailabilityVocabulary` (Elixir) and
/// `shared/availability-vocabulary.json` exactly — the
/// `Saves/SaveVocabulary.swift`/`shared/save-vocabulary.json` contract
/// pattern (D-67), reused here.
///
/// This type exists ONLY so the two codebases are provably proven never
/// to drift apart on the six console filter values' labels and
/// accessible names, via `AvailabilityVocabularyContractTests`. It is
/// NOT the Mac client's own availability state: `AvailabilityState`
/// (six *different* cases, including the storage-view-only
/// `.safeToEvict`) remains the sole read-time derivation this client's
/// own UI ever renders from. No shipped Mac UI reads these constants
/// today — the console filter chips are server-rendered LiveView, not
/// Mac UI — and `shared/availability-vocabulary.json` is a test
/// resource only, never read from a shipped runtime code path.
enum AvailabilityVocabulary {
    static let needsAttention = "needs_attention"
    static let missingDependency = "missing_dependency"
    static let downloading = "downloading"
    static let readyOffline = "ready_offline"
    static let queued = "queued"
    static let serverOnly = "server_only"

    /// The six frozen filter values, in display order — mirrors
    /// `Playstead.AvailabilityVocabulary.values/0` exactly.
    static let values: [String] = [
        needsAttention, missingDependency, downloading, readyOffline, queued, serverOnly
    ]

    static let labels: [String: String] = [
        needsAttention: "Needs attention",
        missingDependency: "Missing dependency",
        downloading: "Downloading",
        readyOffline: "Ready offline",
        queued: "Queued",
        serverOnly: "On server",
    ]

    static let accessibleNames: [String: String] = [
        needsAttention: "Filter by needs attention",
        missingDependency: "Filter by missing dependency",
        downloading: "Filter by downloading",
        readyOffline: "Filter by ready offline",
        queued: "Filter by queued",
        serverOnly: "Filter by on server",
    ]

    /// `label(_:)`, or `nil` if `value` is not in the frozen vocabulary.
    static func label(_ value: String) -> String? { labels[value] }

    /// `accessible_name(_:)`, or `nil` if `value` is not in the frozen
    /// vocabulary.
    static func accessibleName(_ value: String) -> String? { accessibleNames[value] }

    /// `key -> value`, identical in content to
    /// `shared/availability-vocabulary.json`'s flat
    /// `filter.{value}.{label|accessible_name}` keys — the
    /// `SaveVocabulary.all` pattern. The contract test parses the JSON
    /// and asserts this dictionary is exactly equal to it, exhaustively
    /// in both directions.
    static let all: [String: String] = {
        var dict: [String: String] = [:]
        for value in values {
            dict["filter.\(value).label"] = labels[value]
            dict["filter.\(value).accessible_name"] = accessibleNames[value]
        }
        return dict
    }()
}
